source/common/quic/envoy_quic_dispatcher.cc

```diff
@@ -177,13 +177,24 @@ void EnvoyQuicDispatcher::closeConnectionsWithFilterChain(
     // Retain the number of connections in the list early because closing the connection will change
     // the size.
     const size_t num_connections = connections.size();
+    bool delete_sessions_immediately = false;
     for (size_t i = 0; i < num_connections; ++i) {
       Network::Connection& connection = connections.front().get();
       // This will remove the connection from the list. And the last removal will remove connections
       // from the map as well.
       connection.close(Network::ConnectionCloseType::NoFlush);
+      if (!delete_sessions_immediately && dynamic_cast<QuicFilterManagerConnectionImpl&>(connection)
+                                              .extend_filter_manager_lifetime()) {
+        delete_sessions_immediately = true;
```

Comment:
It's a bit surprising that we chose to set "delete immediately" to true when "extend lifetimes" is true. This makes me wonder if this is a bug or if maybe we could add a comment to explain,

Reply:
I was trying to explain it in the comment a few lines below, but probably failed to make it clear. If we want to avoid this explicit deletion, we need to change HCM to have a weak pointer to the HCM config object, because during filter chain config update, the config will be removed before the connection destruction, and since we postponed the filter destruction till the connection destruction, the HCM destructor will crash while accessing config and incrementing stats. The other option is to move the stats incrementation out of HCM destructor. The Envoy defer deletion and the QUIC defer deletion uses different mechanism, so doing any work in the destructor is dangerous.

Reply:
I think it's just the naming.  If it were fix_quic_lifetime_issues() and we deleted FM late and sessions early, both would fit the name.  Deleting sessions based on a variable named extend is a bit weird. If you don't want to rename you could just comment that if this boolean is true runtimeguard X is true and we want to consistently handle quic lifetimes, see below for details?

Reply:
renamed and also added more comment.